src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -0,0 +1,333 @@
+    }
+    else
+    {
+        // Access the light intensity value.
+        size_t item_index = m_nodes[node_index].get_item_index();
+        size_t light_source_index = m_items[item_index].m_light_sources_index;
+        Spectrum spectrum = m_light_sources[light_source_index]->get_intensity();
+
+        // Luminance is the average accross all the spectrum channels
+        for(size_t i = 0; i < spectrum.size(); i++)
+        {
+            luminance += spectrum[i];
+        }
+        luminance /= spectrum.size();
+
+        // Modify the tree depth if the branch is deeper than the other branches
+        // visited so far
+        if (m_tree_depth < node_level)
+            m_tree_depth = node_level;
+    }
+
+    m_nodes[node_index].set_luminance(luminance);
+    m_nodes[node_index].set_level(node_level);
+
+    return luminance;
+}
+
+std::pair<size_t, float> LightTree::sample(
+        const foundation::Vector3d    surface_point,
+        float                         s) const
+{
+    float light_probability = 1.0;
+    size_t node_index = 0;
+
+    while (!m_nodes[node_index].is_leaf())
+    {
+        // LightTreeNode
+        const auto& node = m_nodes[node_index];
+
+        std::pair<float, float> result = child_node_probabilites(node, surface_point);
+        const float p1 = result.first;
+        const float p2 = result.second;
+
+        if (s <= p1)
+        {
+            light_probability *= p1;
+            s /= p1;
+            node_index = node.get_child_node_index();
+        }
+        else
+        {
+            light_probability *= p2;
+            s = (s - p1) / p2;
+            node_index = node.get_child_node_index() + 1;
+        }
+    }
+    size_t item_index = m_nodes[node_index].get_item_index();
+    size_t light_index = m_items[item_index].m_light_tree_external_index;
+
+    return std::pair<size_t, float>(light_index, light_probability);
+}
+
+std::pair<float, float> LightTree::child_node_probabilites(
+        const LightTreeNode<foundation::AABB3d>&    node,
+        const foundation::Vector3d                  surface_point) const
+{
+    const auto& child1 = m_nodes[node.get_child_node_index()];
+    const auto& child2 = m_nodes[node.get_child_node_index() + 1];
+
+    // Node has currently no info about its own bbox characteristics.
+    // Hence we have to extract it before from its parent.
+    // TODO: make LightTreeNode aware of its bbox!
+    const auto& bbox_left  = node.get_left_bbox();
+    const auto& bbox_right = node.get_right_bbox();
+
+    float p1 = node_probability(child1, bbox_left, surface_point);
+    float p2 = node_probability(child2, bbox_right, surface_point);
+
+    // Normalize probabilities
+    float total = p1 + p2;
+    if (total <= 0.0f)
+    {
+        p1 = 0.5;
+        p2 = 0.5;
+    }
+    else
+    {
+        p1 = p1 / total;
+        p2 = p2 / total;
+    }
+
+    return std::pair<float, float>(p1, p2);
+}
+
+float LightTree::node_probability(
```

Comment:
I would suggest making this method a free function inside an anonymous namespace, instead of a class method. It would make it obvious that it's not using anything from the class.